storage/innobase/xtrabackup/src/ds_decrypt.c

```diff
@@ -358,19 +358,23 @@ decrypt_write(ds_file_t *file, const void *buf, size_t len)
 		thd = threads;
 
 		pthread_mutex_lock(&thd->ctrl_mutex);
+		parse_result = thd->parse_result;
 
 		do {
+			size_t new_data_size = 0;
 			if (parse_result == XB_CRYPT_READ_INCOMPLETE) {
-				crypt_file->buf_size = crypt_file->buf_size * 2;
+				crypt_file->buf_size = crypt_file->buf_size + len;
 				crypt_file->buf = (uchar *) my_realloc(
 						crypt_file->buf,
 						crypt_file->buf_size,
 						MYF(MY_FAE|MY_ALLOW_ZERO_PTR));
 			}
 
-			memcpy(crypt_file->buf + crypt_file->buf_len,
-			       buf, MY_MIN(crypt_file->buf_size -
-					   crypt_file->buf_len, len));
+			new_data_size = MY_MIN(crypt_file->buf_size -
+					       crypt_file->buf_len, len);
+			xb_ad(new_data_size > 0);
+			memcpy(crypt_file->buf + crypt_file->buf_len, buf,
+			       new_data_size);
 
 			parse_result = parse_xbcrypt_chunk(
```

Comment:
I think actual bug is here - we need to pass `new_data_size + crypt_file->buf_len` instead of `crypt_file->buf_size`.